### ISSUE TITLE
Only get compiler list once

### ISSUE THREAD

**Issue by wbt** [trufflesuite/truffle#3125]
## Issue
Truffle needlessly downloads the list of Solidity versions both before and after compiling contracts every time.  
For users without a solid Internet connection to that list, this is a more significant holdup and it's not at all necessary.

## Steps to Reproduce
`truffle migrate` on a non-empty project; `truffle compile` would likely work too.
## Actual Results
>Compiling your contracts...
>
>===========================
>× Fetching solc version list from solc-bin. Attempt #1
>× Fetching solc version list from solc-bin. Attempt #2
>[USER'S NOTE: I assume the internet connection to solc-bin is restored right about here.]
>√ Fetching solc version list from solc-bin. Attempt #3
> Compiling <each contract in project>
>√ Fetching solc version list from solc-bin. Attempt #1
> Artifacts written to projectDir\build\contracts
> Compiled successfully using:
>   solc: 0.6.10+commit.00c0fcaf.Emscripten.clang

## Expected Behavior
"Fetching solc version list from solc-bin. Attempt #1..." appears only at the beginning of compilation (ideally, significantly less often than that so that an Internet connection is not required for local offline development.) 

It is not at all clear what benefit there is from fetching the solc version list AFTER compiling the files. 

## Environment
* Operating System: Windows 10
* Truffle version (`truffle version`): v5.1.24
* node version (`node --version`): 9.3.0
* npm version (`npm --version`): 6.4.15

**Comment by lychees** [comment]
Seconded.

**Comment by haltman-at** [comment]
Thanks for pointing this out!  We'll take a look.

**Comment by eggplantzzz** [comment]
@wbt Thanks for the report! I don't seem to be able to reproduce your logging. Can you confirm that you get the same results on the latest Truffle (5.1.33)?

**Comment by wbt** [comment]
@eggplantzzz Yes, I do see it (as before, not always with failures to reach solc-bin).
When running `truffle compile` I get this output:

>Compiling your contracts...
===========================
√ Fetching solc version list from solc-bin. Attempt #1
> Compiling [each .sol file in .\contracts or used from npm]
√ Fetching solc version list from solc-bin. Attempt #1
> Artifacts written to myProjectDir\build\contracts
> Compiled successfully using:
>   \- solc: 0.6.10+commit.00c0fcaf.Emscripten.clang

You can see the `√ Fetching solc version list from solc-bin. Attempt #1` line repeated.

**Comment by eggplantzzz** [comment]
Ok, thanks very much for that @wbt. I'll try and see if I can find some time to look into this!

**Comment by wbt** [comment]
I suspect the cause is an extra call to [`getSolcVersions()`](https://github.com/trufflesuite/truffle/blob/2a5b011cd859b9a2e5ad1bf4eed28d057050cf06/packages/compile-solidity/compilerSupplier/loadingStrategies/VersionRange.js#L141) somewhere. I found four calls in the codebase:

- In [getSolcFromCacheOrUrl](https://github.com/trufflesuite/truffle/blob/2a5b011cd859b9a2e5ad1bf4eed28d057050cf06/packages/compile-solidity/compilerSupplier/loadingStrategies/VersionRange.js#L88) which is called in [VersionRange.load](https://github.com/trufflesuite/truffle/blob/2a5b011cd859b9a2e5ad1bf4eed28d057050cf06/packages/compile-solidity/compilerSupplier/loadingStrategies/VersionRange.js#L191), [downloadAndCacheSolc](https://github.com/trufflesuite/truffle/blob/f3804c2c36f993317c0d23202e09726a9dd6222c/packages/compile-solidity/compilerSupplier/index.js#L34), and [this test file](https://github.com/trufflesuite/truffle/blob/develop/packages/compile-solidity/test/compilerSupplier/loadingStrategies/VersionRange.js).

- In [getSolcByCommit](https://github.com/trufflesuite/truffle/blob/2a5b011cd859b9a2e5ad1bf4eed28d057050cf06/packages/compile-solidity/compilerSupplier/loadingStrategies/VersionRange.js#L88), which on a quick search appears to only be called in [this test file](https://github.com/trufflesuite/truffle/blob/develop/packages/compile-solidity/test/compilerSupplier/loadingStrategies/VersionRange.js).

- In [getReleases](https://github.com/trufflesuite/truffle/blob/f3804c2c36f993317c0d23202e09726a9dd6222c/packages/compile-solidity/compilerSupplier/index.js#L105) which on a quick search appears to be called just once [in listVersions](https://github.com/trufflesuite/truffle/blob/2777307c5c78c4d8f6ac5f3c15bf15e371074176/packages/core/lib/commands/compile.js#L88) called just [here](https://github.com/trufflesuite/truffle/blob/2777307c5c78c4d8f6ac5f3c15bf15e371074176/packages/core/lib/commands/compile.js#L58) for handling the `--list` option to `truffle compile`.

- In [this test](https://github.com/trufflesuite/truffle/blob/42e2511f93203e62920c41519decd68a1529ac63/packages/compile-solidity/test/compilerSupplier/loadingStrategies/VersionRange.js#L84). 

Diving into more detail on the first one:

downloadAndCacheSolc is called in [obtain.js:run()](https://github.com/trufflesuite/truffle/blob/42e2511f93203e62920c41519decd68a1529ac63/packages/core/lib/commands/obtain.js#L28) and [this test file](https://github.com/trufflesuite/truffle/blob/develop/packages/core/test/lib/commands/obtain.js).

VersionRange.load is called in at least [loadParserSolc](https://github.com/trufflesuite/truffle/blob/f3804c2c36f993317c0d23202e09726a9dd6222c/packages/compile-solidity/compilerSupplier/index.js#L77) and [this test file](https://github.com/trufflesuite/truffle/blob/develop/packages/compile-solidity/test/compilerSupplier/index.js). loadParserSolc is [called](https://github.com/trufflesuite/truffle/blob/f3804c2c36f993317c0d23202e09726a9dd6222c/packages/compile-solidity/compilerSupplier/index.js#L70) in CompilerSupplier.load, which is called from at least [compile-solidity/profiler/index.js](https://github.com/trufflesuite/truffle/blob/1ead25b2ec88002639354cad8a88046830e32bd8/packages/compile-solidity/profiler/index.js#L98) and [compile-solidity/run.js's invokeCompiler()](https://github.com/trufflesuite/truffle/blob/22401cecd4cc194b7a3de3bbda5041dbf98f9fa3/packages/compile-solidity/run.js#L252).

Then I ran out of time for tracing through.  Hope this helps!

**Comment by wbt** [comment]
Here's a brief walkthrough:

From [compile.with_dependencies](https://github.com/trufflesuite/truffle/blob/5e3dd1d72ff920e221c32fade4530f8cbcbce2e1/packages/compile-solidity/new/index.js#L69), there is a [call](https://github.com/trufflesuite/truffle/blob/5e3dd1d72ff920e221c32fade4530f8cbcbce2e1/packages/compile-solidity/new/index.js#L83) to [Profiler.required_sources](https://github.com/trufflesuite/truffle/blob/1ead25b2ec88002639354cad8a88046830e32bd8/packages/compile-solidity/profiler/index.js#L65), which [calls](https://github.com/trufflesuite/truffle/blob/1ead25b2ec88002639354cad8a88046830e32bd8/packages/compile-solidity/profiler/index.js#L98) CompilerSupplier.load().

Later in compile.with_dependencies, there is a [call](https://github.com/trufflesuite/truffle/blob/5e3dd1d72ff920e221c32fade4530f8cbcbce2e1/packages/compile-solidity/new/index.js#L106) to [compile](https://github.com/trufflesuite/truffle/blob/5e3dd1d72ff920e221c32fade4530f8cbcbce2e1/packages/compile-solidity/new/index.js#L22) which calls [run](https://github.com/trufflesuite/truffle/blob/22401cecd4cc194b7a3de3bbda5041dbf98f9fa3/packages/compile-solidity/run.js#L8) which [calls](https://github.com/trufflesuite/truffle/blob/22401cecd4cc194b7a3de3bbda5041dbf98f9fa3/packages/compile-solidity/run.js#L29) [invokeCompiler](https://github.com/trufflesuite/truffle/blob/22401cecd4cc194b7a3de3bbda5041dbf98f9fa3/packages/compile-solidity/run.js#L245) which **also** creates a new CompilerSupplier and [calls](https://github.com/trufflesuite/truffle/blob/22401cecd4cc194b7a3de3bbda5041dbf98f9fa3/packages/compile-solidity/run.js#L252) load() on it. 

The trace from CompilerSupplier.load to getSolcVersions() is in my immediately previous post in this thread.

It is important to note that when Truffle says "Compiling [file]" it isn't actually compiling at that time, it's just using poorly-chosen phrasing to say that it will at some point in the near future be compiling that file.  ("Staging [filename] for compilation" would be clearer.) This has previously been a source of frustrating confusion and remains so. Thus, the second time it's loading the list of compilers, it still hasn't actually done any compilation, so that second time isn't happening after compilation as I reasonably thought when writing the OP above.

I would thus revise my suggestion to indicate that the **first** loading of the compiler version list is probably the unnecessary one which can be most easily cut.  I would still like to see more caching to support totally offline development, but hopefully this can give a strong pointer to where the redundancy lies.

**Comment by akutruff** [comment]
Any progress or way to force caching here?  This is the slowest part of my dev loop

**Comment by Aeshah1122** [comment]
how can solution of this problim? plz

**Comment by eggplantzzz** [comment]
@akutruff Are you also using Windows? It is in the realm of possibilities that this is Windows-related. I don't seem to be able to reproduce this issue on a mac. Once you download a compiler, Truffle caches it by default.

**Comment by wbt** [comment]
While I am using it on Windows, I think the code trace above indicates a problem there.

**Comment by akutruff** [comment]
@eggplantzzz Im running in an Ubuntu docker container.  It is a container inside WSL2, but that shouldn't matter, right?  The container is using a true docker volume for the code/node_modules and the OS and is not a manual mount to a windows directory.

**Comment by eggplantzzz** [comment]
I'm not sure if it matters or not, I am unable to duplicate the behavior on my machine that is described by the OP. Very strange to me...

**Comment by wbt** [comment]
You don't see a sequence like this?: 

>√ Fetching solc version list from solc-bin. Attempt #1
Compiling [each contract in project]
√ Fetching solc version list from solc-bin. Attempt #1

**Comment by eggplantzzz** [comment]
No, if I don't have the compiler it fetches it but only once for me.

**Comment by xcsean** [comment]
check your truffle-config.js and see if the solc version set to "^x.x.x", if you use "^" that means not sure which version of solc is use, then truffle will check&get solc every time.

**Comment by wbt** [comment]
@xcsean Fine if it checks every time, but ONCE per compile should be enough.

**Comment by eggplantzzz** [comment]
Seems like the simplest solution to this is memoization inside `CompilerSupplier`. Just want to note that there is an edge case with the `compilers.solc.parser` option designed to allow users to parse with soljson even though they specify compilation with docker. See [this](https://trufflesuite.com/docs/truffle/reference/configuration/#solc) for some more info.
